### PR TITLE
Add QGaussRadau quadrature formula + tests

### PR DESCRIPTION
**Pull request by jpthiele**
This adds the Gauss-Radau quadrature formulae up to n=3 points,
as described here: https://mathworld.wolfram.com/RadauQuadrature.html.
The given numerical values for n=4 and n=5 are not accurate enough
and application to the test polynomials
of quadrature_test.cc were not exact for degrees > 0.

On the 3d faces and subfaces the QF is able to integrate the given
polynomial in the test suite of one order higher than in all other cases.
This is reflected in the quadrature_test.ouput file for numdiff comparison.

### CODE REVIEW

**Comment by jpthiele** [comment]
Addendum: As the right Gauss-Radau formula quadrature points are obtained by subtracting the left quadrature points from 1,
the resulting formula is not sorted. This results in some additional output in the test file. 
I asked @kronbichler today whether the points should be reordered and he said that it should not matter to his knowledge,
so I opted for skipping a sorting step.

File: include/deal.II/base/quadrature_lib.h
Comment (peterrum) [inline include/deal.II/base/quadrature_lib.h:79]:
```suggestion
  enum class EndPoint
```

File: include/deal.II/base/quadrature_lib.h
Comment (peterrum) [inline include/deal.II/base/quadrature_lib.h:99]:
```suggestion
  const EndPoint end_point;
```

File: include/deal.II/base/quadrature_lib.h
Comment (peterrum) [inline include/deal.II/base/quadrature_lib.h:51]: Could you add a sentence about when this quadrature rule is useful.

File: source/base/quadrature_lib.cc
Comment (peterrum) [inline source/base/quadrature_lib.cc:192]: Personally, I would prefer if the points were sorted. Could you flip the direction of the for loop?

File: source/base/quadrature_lib.cc
Comment (peterrum) [inline source/base/quadrature_lib.cc:208]:
```suggestion
    get_left_quadrature_weights(const unsigned int n)
```

**Review by peterrum — COMMENTED** [review]
Looks overall good. I have some minor comments. Could you add a change-log entry?

File: source/base/quadrature_lib.cc
Comment (peterrum) [inline source/base/quadrature_lib.cc:201]:
```suggestion
          return {};
        }
```

File: include/deal.II/base/quadrature_lib.h
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:99]: This is the same notation as in QGaussRadauChebychev,
but I'd be open to change both in this or a second PR.

Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:102]: The notation here is the same as in QGaussRadauChebychev,
so I opted for consistency.

File: include/deal.II/base/quadrature_lib.h
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:51]: Done

File: source/base/quadrature_lib.cc
Comment (jpthiele) [reply source/base/quadrature_lib.cc]: Done

File: tests/base/quadrature_test.output
Comment (peterrum) [inline tests/base/quadrature_test.output:143]: @jpthiele  Looks correct now!?

Comment (jpthiele) [reply tests/base/quadrature_test.output:143]: Yes, the quadrature formulas are indeed correct now.
Interestingly the final few digits seem to have had quite a large impact.

**Review by peterrum — APPROVED** [review]

File: include/deal.II/base/quadrature_lib.h
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:55]: degrees -> degree

File: include/deal.II/base/quadrature_lib.h
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:61]: This works, but you can also use $latex-style$ formulas if you think that would be easier to read.

File: include/deal.II/base/quadrature_lib.h
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:73]:
```suggestion
 * @seealso https://mathworld.wolfram.com/RadauQuadrature.html
```

File: include/deal.II/base/quadrature_lib.h
Comment (bangerth) [reply include/deal.II/base/quadrature_lib.h:102]: That would be great! Feel free to leave it to a follow-up patch!

File: include/deal.II/base/quadrature_lib.h
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:1001]:
```suggestion
QGaussRadau<1>::QGaussRadau(const unsigned int n, const QGaussRadau<1>::EndPoint ep);
```

File: source/base/quadrature_lib.cc
Comment (bangerth) [inline source/base/quadrature_lib.cc:220]:
```suggestion
            Assert(false, ExcNotImplemented());
```

File: source/base/quadrature_lib.cc
Comment (bangerth) [inline source/base/quadrature_lib.cc:360]:
```suggestion
  Assert(n > 0, ExcMessage("Need at least one point for quadrature rules."));
```

**Review by bangerth — COMMENTED** [review]
Very nice! Some comments, none should be much work!

**Comment by kronbichler** [comment]
To help me understand, I'm still wondering why the test previously passed for the left end point but not the right end point. It must be that we do not test the complete polynomial basis but only the monomials via https://github.com/dealii/dealii/blob/e8073dd8fabb7ec9b982aefe4efff7a0851e7782/tests/base/quadrature_test.cc#L78-L87 While the monomials form a basis, they form a pretty ill-conditioned one, so we might probably want to also look at `(1-x)` or `(0.5 - x)` for comparison just to be safe in terms of roundoff. If you agree @jpthiele I can post a follow-up later.

**Comment by jpthiele** [comment]
> To help me understand, I'm still wondering why the test previously passed for the left end point but not the right end point. 

In contrast to many other quadrature formulae the Gauss-Radau family is not symmetrical around 0 for [-1,1].
We did temporarily change the 1d test from x^i to (1-x)^i and the behaviour switched,
such that the left formula failed and the right formula passed.

@mathmerizing found a Julia package that computes Gauss formula of different families 
and it gave weights that differed to the previous results in the last three significant digits (i.e. around 14 identical ones).
There are two formulae listed for the weights 
and maybe the one with the derivative of the Legendre polynomials is a bit more accurate, 
if that was the one used in the other package.

File: include/deal.II/base/quadrature_lib.h
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:102]: As there are various comments concerning the structure and commenting of both QGaussRadau 
and QGaussRadauChebychev I decided to apply all comments to QGaussRadau here.
After merging I will go over the QGaussRadauChebychev class 
and make all applicable changes to unify documentation and interface.

File: include/deal.II/base/quadrature_lib.h
Comment (jpthiele) [inline include/deal.II/base/quadrature_lib.h:68]: should be 

...\text{and}\quad
w_i=...

will fix that tomorrow.

File: include/deal.II/base/quadrature_lib.h
Comment (jpthiele) [inline include/deal.II/base/quadrature_lib.h:72]: Change to
... the quadrature points are $...$ and the weights are $\tilde{w}_i = w_{n-i-1}$, 
with $(x_i,w_i)$ as quadrature points and weights of the left Gauss-Radau formula.

File: include/deal.II/base/quadrature_lib.h
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h]: Yes it is easier to read. Changed it, thanks!

File: source/base/quadrature_lib.cc
Comment (kronbichler) [inline source/base/quadrature_lib.cc:241]: Can you please move the declaration of `points` in here, i.e.,
```suggestion
          case ::dealii::QGaussRadau<1>::EndPoint::right:
            {
              std::vector<double> points(n);
              for (unsigned int i = 0; i < n; ++i)
                {
                  points[n - i - 1] = 1. - left_points[i];
                }
              return points;
            }
```
We aim to have the declaration near the place of use. As a side product, you avoid setting up the array when it's not needed.

**Review by kronbichler — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by bangerth** [comment]
Yay, welcome to the authors of deal.II!

**Pull request merged**
